backend/src/main/scala/bloop/reporter/Reporter.scala

```diff
@@ -44,13 +44,39 @@ abstract class Reporter(
     ()
   }
 
-  override def hasErrors(): Boolean = hasErrors(_problems)
-  override def hasWarnings(): Boolean = hasWarnings(_problems)
+  override def hasWarnings(): Boolean = {
+    _problems.exists(_.problem.severity == Severity.Warn)
+  }
+
+  override def hasErrors(): Boolean = {
+    _problems.exists(p => p.problem.severity == Severity.Error)
+  }
 
   override def problems(): Array[xsbti.Problem] = _problems.map(_.problem).toArray
   override def allProblems: Seq[Problem] = _problems.map(p => liftProblem(p.problem)).toList
   override def allProblemsPerPhase: Seq[ProblemPerPhase] = _problems.toList
 
+  private var hasFatalWarningsEnabled: Boolean = false
+  override def enableFatalWarnings(): Unit = {
+    hasFatalWarningsEnabled = true
+  }
+
+  private val sourceFilesWithFatalWarnings = TrieMap.empty[File, Boolean]
+  override def getSourceFilesWithFatalWarnings: Set[File] = {
+    sourceFilesWithFatalWarnings.keysIterator.toSet
+  }
+
+  protected def liftFatalWarning(problem: Problem): Problem = {
+    if (!hasFatalWarningsEnabled || problem.severity != Severity.Warn) problem
```

Comment:
I find not-negated conditions easier to reason about:
`!hasFatalWarningsEnabled || problem.severity != Severity.Warn`
vs
`hasFatalWarningsEnabled && problem.severity == Severity.Warn`

What do you think?

Reply:
Wanted to suggest the same thing. It's the prefect De Morgan's law example :P

Reply:
I can hoist up this value to another variable and then negate it. I always like to handle the happy path in the first if branch and then handle the complicated case in the else branch, that's why I'm using here a negated condition.